torch_agents/cogment_verse_torch_agents/muzero/adapter.py

```diff
@@ -0,0 +1,801 @@
+        reward_distribution,
+        value_distribution,
+    )
+    reanalyze_agent_queue = [mp.Queue() for _ in range(num_reanalyze_workers)]
+    reanalyze_workers = [
+        ReanalyzeWorker(
+            reanalyze_queue,
+            reanalyze_update_queue,
+            reanalyze_agent_queue[i],
+            config.training.reanalyze_device,
+            reward_distribution,
+            value_distribution,
+        )
+        for i in range(num_reanalyze_workers)
+    ]
+
+    trials_completed = 0
+    running_stats = RunningStats()
+
+    xp_tracker.log_params(
+        config.environment,
+        config.actor,
+        config.training,
+    )
+
+    trial_configs = make_trial_configs(run_session, config, model_id, -1)
+
+    total_samples = 0
+    training_step = 0
+    run_total_reward = 0
+
+    workers = [train_worker, replay_buffer] + reanalyze_workers
+
+    try:
+        for worker in workers:
+            worker.start()
+
+        start_time = time.time()
+
+        sample_generator = run_session.start_trials_and_wait_for_termination(
+            trial_configs,
+            max_parallel_trials=config.training.max_parallel_trials,
+        )
+
+        start_time = time.time()
+        samples = 0
+
+        async for _step, timestamp, trial_id, _tick, (sample, total_reward) in sample_generator:
+            samples += 1
+            replay_buffer.add_sample(trial_id, sample)
+            total_samples += 1
+
+            for worker in workers:
+                assert worker.is_alive()
+
+            if sample.done:
+                trials_completed += 1
+                xp_tracker.log_metrics(
+                    timestamp, total_samples, trial_total_reward=total_reward, trials_completed=trials_completed
+                )
+                run_total_reward += total_reward
+
+            if replay_buffer.size() <= config.training.min_replay_buffer_size:
+                continue
+
+            # get latest online model
+            try:
+                agent = agent_update_queue.get_nowait()
+                cpu_agent = copy.deepcopy(agent)
+                cpu_agent.set_device("cpu")
+                for agent_queue in reanalyze_agent_queue:
+                    agent_queue.put(cpu_agent)
+            except queue.Empty:
+                pass
+
+            while not info_queue.empty():
+                try:
+                    info = info_queue.get_nowait()
+                except queue.Empty:
+                    continue
+
+                training_step += 1
+                info["model_version"] = version_info["version_number"]
+                info["training_step"] = training_step
+                info["mean_trial_reward"] = run_total_reward / max(1, trials_completed)
+                info["samples_per_sec"] = total_samples / max(1, time.time() - start_time)
+                info["reanalyzed_samples"] = sum([worker.reanalyzed_samples() for worker in reanalyze_workers])
+                running_stats.update(info)
```

Comment:
It seems to me this mechanism shouldn't be needed, can't we just deal with that in the tracker ?

Reply:
let's push that to when we have to solve it again.